daceml/autodiff/utils.py

```diff
@@ -0,0 +1,189 @@
+                    name, forward_node.schema.name))
+
+    program.__annotations__ = {**inputs, **outputs}
+
+    sdfg = DaceProgram(program, (), {}).to_sdfg()
+
+    result_node = context.backward_state.add_nested_sdfg(
+        sdfg, None, set(inputs), set(outputs))
+
+    return result_node, backward_result
+
+
+def connect_output_from_forward(forward_node: nd.Node, backward_node: nd.Node,
+                                context: BackwardContext,
+                                output_connector_name: str):
+    """ Connect an output of the forward node as an input to the backward node. This is done by forwarding the array
+        from the forward pass.
+
+        Conceptually, this is similar to pytorch's ctx.save_for_backward.
+
+        :param forward_node: the node in the forward pass.
+        :param backward_node: the node in the backward pass.
+        :param context: the backward context.
+        :param output_connector_name: the name of the connector on the backward pass. The output of that connector will
+                                      be forwarded to the connector of the same name on the backward node.
+    """
+    output_edge = utils.out_edge_with_name(forward_node, context.forward_state,
+                                           output_connector_name)
+
+    # add the array of the output to backward_input_arrays that it will be forwarded by the autodiff engine
+    output_arr_name = output_edge.data.data
+    if output_arr_name not in context.backward_generator.backward_input_arrays:
+        data_desc = context.forward_sdfg.arrays[output_arr_name]
+        context.backward_generator.backward_input_arrays[
+            output_arr_name] = copy.deepcopy(data_desc)
+
+        if context.backward_generator.separate_sdfgs:
+            data_desc.transient = False
+            context.backward_sdfg.add_datadesc(output_arr_name, data_desc)
+
+        read = context.backward_state.add_read(output_arr_name)
+    else:
+        cand = [
+            n for n, _ in context.backward_state.all_nodes_recursive()
+            if isinstance(n, nd.AccessNode) and n.data == output_arr_name
+        ]
+        assert len(cand) == 1
+        read = cand[0]
+    context.backward_state.add_edge(read, None, backward_node,
+                                    output_connector_name,
+                                    copy.deepcopy(output_edge.data))
+
+
+def cast_consts_to_type(code: str, dtype: dace.typeclass) -> str:
+    """ Convert a piece of code so that constants are wrapped in casts to ``dtype``.
+
+        For example:
+
+            x * ( 3 / 2)
+
+        becomes:
+
+            x * (dace.float32(3) / dace.float32(2))
+
+        :param code: the code string to convert.
+        :param dtype: the dace typeclass to wrap cast to
+        :return: a string of the converted code.
+    """
+    class CastConsts(ast.NodeTransformer):
+        def visit_Num(self, node):
+            return ast.parse(
```

Comment:
needs `ast.copy_location` of the old node, and also `.body[0].value` I think (the output is an ast.Module, and you can't put this in a valid piece of code (will also generate extraneous newlines)